### PR TITLE
[FSE - Templates]: Add info to templates created from wp suggestion

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Part of: https://github.com/WordPress/gutenberg/issues/37407
Needed for: https://github.com/WordPress/gutenberg/pull/41189

This PR adds a post meta(`is_wp_suggestion`) to templates created from the site editor.

## Why?
Currently we handle some template actions/filtering based on whether the template is part of the template hierarchy. This currently happens through `get_default_block_template_types` results, that are hardcoded and of course are just a few of the possible templates. Many specific templates will need to be dynamically calculated and suggested(see linked PR for single-$post_type, archive-$post_type and post specific templates) and these should be treated as well as part of the template hierarchy.

In a few words we need to differentiate the templates created from the post editor in the `Template` panel in inspector controls and the ones suggested in site editor. So in the suggested templates in site editor list(`Add new`) from WP we should mark them as suggested.

## Testing instructions
The two main differences are:
1. in site editor we should not be able to rename a template from template hierarchy
2. such a template should not be shown in post editor available templates
3. Create a new template from post editor
4. Add the below snippet [here(before the return)](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/add-new-template/new-template.js#L154) and create the template
`missingTemplates.push( {
	title: 'Suggested',
	description: 'test template',
	slug: 'single-book',
} );

3. Test that everything is working as expected and no regression is introduced
6. Existing templates should rename unaffected.

### CODE REVIEW

File: lib/compat/wordpress-6.1/class-gutenberg-rest-templates-controller.php [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		} elseif ( null !== $template && 'custom' !== $template->source ) {
+			$changes->post_title = $template->title;
+		}
+		if ( isset( $request['description'] ) ) {
+			$changes->post_excerpt = $request['description'];
+		} elseif ( null !== $template && 'custom' !== $template->source ) {
+			$changes->post_excerpt = $template->description;
+		}
+
+		if ( 'wp_template' === $this->post_type ) {
```
Comment (ntsekouras) [inline lib/compat/wordpress-6.1/class-gutenberg-rest-templates-controller.php:144]: This is a new addition.

File: lib/compat/wordpress-6.1/block-template-utils.php [hunk 2]
```diff
@@ -27,3 +27,267 @@ function gutenberg_get_default_block_template_types( $default_template_types ) {
...
+	if ( is_wp_error( $terms ) ) {
+		return $terms;
+	}
+
+	if ( ! $terms ) {
+		return new WP_Error( 'template_missing_theme', __( 'No theme is defined for this template.', 'gutenberg' ) );
+	}
+
+	$origin           = get_post_meta( $post->ID, 'origin', true );
+	$is_wp_suggestion = get_post_meta( $post->ID, 'is_wp_suggestion', true );
```
Comment (ntsekouras) [inline lib/compat/wordpress-6.1/block-template-utils.php:258]: This is a new addition.

File: lib/compat/wordpress-6.1/block-template-utils.php [hunk 3]
```diff
@@ -27,3 +27,267 @@ function gutenberg_get_default_block_template_types( $default_template_types ) {
...
+	 *
+	 *     @type array  $slug__in List of slugs to include.
+	 *     @type int    $wp_id Post ID of customized template.
+	 *     @type string $post_type Post type to get the templates for.
+	 * }
+	 * @param array $template_type wp_template or wp_template_part.
+	 */
+	$templates = apply_filters( 'pre_get_block_templates', null, $query, $template_type );
+	if ( ! is_null( $templates ) ) {
+		echo 'hi';
```
Comment (Mamaduka) [inline lib/compat/wordpress-6.1/block-template-utils.php]: Leftovers :)

File: lib/compat/wordpress-6.1/block-template-utils.php [hunk 4]
```diff
@@ -27,3 +27,267 @@ function gutenberg_get_default_block_template_types( $default_template_types ) {
...
+	/**
+	 * Filters the queried block template object after it's been fetched.
+	 *
+	 * @since 10.8
+	 *
+	 * @param Gutenberg_Block_Template|null $block_template The found block template, or null if there isn't one.
+	 * @param string $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array  $template_type wp_template or wp_template_part.
+	 */
+	return apply_filters( 'get_block_template', $block_template, $id, $template_type );
```
Comment (Mamaduka) [inline lib/compat/wordpress-6.1/block-template-utils.php:235]: Unrelated to this PR, but this filter will never apply if we have template object results from DB. See `is_wp_error` condition above.

File: lib/compat/wordpress-6.1/block-template-utils.php (hunk 3 above)
Comment (ntsekouras) [reply lib/compat/wordpress-6.1/block-template-utils.php]: haha, sorry about that 😄

File: lib/compat/wordpress-6.1/block-template-utils.php (hunk 4 above)
Comment (ntsekouras) [reply lib/compat/wordpress-6.1/block-template-utils.php:235]: I noticed that as well, but it might be better to handle this separately. WDYT?

Comment (ntsekouras) [reply lib/compat/wordpress-6.1/block-template-utils.php:235]: Also it might be intentional, I'm not really sure and would need checking.

Comment (Mamaduka) [reply lib/compat/wordpress-6.1/block-template-utils.php:235]: Sure. Let's fix this in a follow-up.

**Review by Mamaduka — APPROVED** [review]
Changes test well for me. Thanks, Nik!

✅  I cannot rename templates from the WP hierarchy in Site Editor.
✅  These templates aren't visible in the Post Editor template dropdown (The default template doesn't count).
✅ Suggested template create works as expected, and they're not visible in the post editor.

If not for the issue I [mentioned above](https://github.com/WordPress/gutenberg/pull/41387#discussion_r883379120), I think we could've done most of the changes using `get_block_templates` and `get_block_template` filters + extending the REST API endpoint.

The number of changes we've to make for simple enhancements like this sure makes things complicated.

**Comment by DeoThemes** [comment]
Does this commit have something to do with the recent issue when the page templates are no longer displaying? - https://github.com/WordPress/gutenberg/issues/41617

**Comment by ntsekouras** [comment]
>Does this commit have something to do with the recent issue when the page templates are no longer displaying? - https://github.com/WordPress/gutenberg/issues/41617

It seems it does. I'll look into this!

**Comment by mburridge** [comment]
Added the Needs Dev Note label in case this needs a dev note for the 6.1 release.

**Pull request merged**